awx/main/fields.py

```diff
@@ -197,18 +199,28 @@ def update_role_parentage_for_instance(instance):
     updates the parents listing for all the roles
     of a given instance if they have changed
     '''
+    changed_ct = 0
+    parents_removed = set()
+    parents_added = set()
     for implicit_role_field in getattr(instance.__class__, '__implicit_role_fields'):
         cur_role = getattr(instance, implicit_role_field.name)
         original_parents = set(json.loads(cur_role.implicit_parents))
         new_parents = implicit_role_field._resolve_parent_roles(instance)
-        cur_role.parents.remove(*list(original_parents - new_parents))
-        cur_role.parents.add(*list(new_parents - original_parents))
+        removals = original_parents - new_parents
+        if removals:
+            cur_role.parents.remove(*list(removals))
+            parents_removed.add(cur_role.pk)
+        additions = new_parents - original_parents
+        if additions:
+            cur_role.parents.add(*list(additions))
+            parents_added.add(cur_role.pk)
         new_parents_list = list(new_parents)
         new_parents_list.sort()
         new_parents_json = json.dumps(new_parents_list)
         if cur_role.implicit_parents != new_parents_json:
             cur_role.implicit_parents = new_parents_json
-            cur_role.save()
+            cur_role.save(update_fields=['implicit_parents'])
+    return (parents_added, parents_removed)
```

Comment:
I understand that this probably doesn't look good. I have tried reverting the changes in this file, and it doesn't work, and I am reminded of why I made the changes.

When the existing method in migrations

https://github.com/ansible/awx/blob/c626f51dae09fb7770bc313f7d7f28e86ef2b9a3/awx/main/migrations/_rbac.py#L38-L50

is called, it only passes in _additions_ and not _removals_. Normally, this is called within signals, where additions and removals are passed on based on what related manager is changed.

Without passing in removals, I can't use that. Well, that's also the only thing that we have for this sort of thing in migration code. If you pull in current models, it doesn't even come close to working because event the schema changes in this PR break the signals / fields logic for implicit role fields. Even if they didn't break here, they would break with future changes.

Yes, it is _extremely_ concerning that our migrations don't seem to re-cache ancestors for removals of permissions. But that method is also probably razor close to OOMing with real databases, and passes around a massive payload in the SQL. Putting in removals of all root nodes to that is also a recipe for breaking things.

So, feeling kind of cornered, I feel like the only logical thing to do is follow the pattern that the signals are following.